next/src/visibility.ts

```diff
@@ -21,41 +33,59 @@ export function updateFieldVisibility(
     return
   }
 
+  // Reseting fields visibility to the default before re-calculating it
+  resetVisibility(fields)
+
   // Apply rules to current level of fields
   applySchemaRules(fields, values, schema, options)
 
   // Process nested object fields that have conditional logic
   for (const fieldName in schema.properties) {
     const fieldSchema = schema.properties[fieldName]
+    const field = fields.find(field => field.name === fieldName)
 
-    // Only process object schemas with conditional logic (allOf)
-    if (typeof fieldSchema !== 'object' || fieldSchema === null
-      || Array.isArray(fieldSchema) || !fieldSchema.allOf) {
-      continue
-    }
-
-    const objectField = fields.find(field => field.name === fieldSchema.title || field.name === fieldName)
-    if (!objectField || !objectField.fields || objectField.fields.length === 0) {
-      continue
+    if (field?.fields) {
+      applySchemaRules(field.fields, values[fieldName], fieldSchema as JsfObjectSchema, options)
     }
+  }
+}
 
-    const fieldValues = isObjectValue(values[fieldName])
-      ? values[fieldName]
-      : isObjectValue(values[objectField.name]) ? values[objectField.name] : {}
-
-    // Apply rules to nested fields
-    applySchemaRules(objectField.fields, fieldValues, fieldSchema as JsfObjectSchema, options)
-
-    // Only process nested fields if parent is visible
-    if (objectField.isVisible) {
-      updateFieldVisibility(
-        objectField.fields,
-        fieldValues,
-        fieldSchema as JsfObjectSchema,
-        options,
-      )
-    }
+/**
+ * Evaluates the conditional rules for a field
+ * @param values - The current field values
+ * @param schema - The JSON schema definition
+ * @param rule - Schema identifying the conditional rule
+ * @param options - Validation options
+ * @returns An object containing the rule and whether it matches
+ */
+function evaluateConditional(
+  values: ObjectValue,
+  schema: JsfObjectSchema,
+  rule: NonBooleanJsfSchema,
+  options: ValidationOptions = {},
+) {
+  const ifErrors = validateSchema(values, rule.if!, options)
+  const matches = ifErrors.length === 0
+
+  // Prevent fields from being shown when required fields have type errors
+  let hasTypeErrors = false
+  if (matches
+    && typeof rule.if === 'object'
+    && rule.if !== null
+    && Array.isArray(rule.if.required)) {
+    const requiredFields = rule.if.required
```

Comment:
bug(ts): We forgot to type `required` here. And I think this can be simplified to:

```js
if (matches && Array.isArray(rule.if?.required)) {
```

Reply:
I don't know if you did this to have defensive code, but `if.required` is _always_ an array, so the `isArray` is redundant.

Reply:
Can I improve this in a separate PR? I need to fight with types and I want to do it step by step

Reply:
ofc, approved in the meantime :)